### PR TITLE
Panic at search when the trees need a rebuild

### LINKED ISSUE #74: Reader do not detect that the database has been partially updated
We need to do an investigation

### PR DESCRIPTION
**Pull request by irevoire**
# Pull Request

## Related issue
Fixes #74 

## What does this PR do?
- Ensure the database has been fully built before attempting any search request (=> there are no items to update)
- Delete the `updated` item list instead of emptying it
- Provide a new method on the `Writer` to know if we need to build the tree or can simply drop the writer
- Add a new test
- Return an error message instead of panicking when a key is missing

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -40,10 +42,16 @@ pub enum Error {
...
+
+    /// The last time items in the database were updated, the [`Writer::build`] method wasn't called.
+    #[error("The trees have not been built after an update on index {0}.")]
+    NeedBuild(u16),
 
     /// Internal error
-    #[error("Internal error: Node is missing")]
-    MissingNode,
+    #[error("Internal error: {:?} is missing in index `{}`.", .0.node, .0.index)]
+    MissingKey(Key),
```
Comment (Kerollmops) [inline src/error.rs:56]: This new error variant is public. Is the `Key` type public, or should we split the fields and not expose this type?

File: src/error.rs [hunk 2]
```diff
@@ -40,10 +42,16 @@ pub enum Error {
 
     /// Arroy is not able to find the metadata for a given index.
     /// It is probably because the user forget to build the database.
-    #[error("Metadata are missing, did you build your database before trying to read it.")]
-    MissingMetadata,
+    #[error(
+        "Metadata are missing on index {0}, did you build your database before trying to read it."
```
Comment (Kerollmops) [inline src/error.rs:46]:
```suggestion
        "Metadata are missing on index {0}. You must build your database before attempting to read it."
```

File: src/reader.rs [hunk 3]
```diff
@@ -45,6 +45,11 @@ impl<'t, D: Distance> Reader<'t, D> {
                 received: D::name(),
             });
         }
+        let need_build =
+            database.remap_data_type::<Bytes>().get(rtxn, &Key::updated(index))?.is_some();
+        if need_build {
```
Comment (Kerollmops) [inline src/reader.rs:50]:
```suggestion
        if database.remap_data_type::<Bytes>().get(rtxn, &Key::updated(index))?.is_some() {
```

File: src/tests/writer.rs [hunk 4]
```diff
@@ -92,12 +90,12 @@ fn write_one_vector() {
     Item 0: Leaf(Leaf { header: NodeHeaderEuclidean { bias: 0.0 }, vector: [0.0000, 1.0000, 2.0000] })
     Tree 0: Descendants(Descendants { descendants: [0] })
     Root: Metadata { dimensions: 3, items: RoaringBitmap<[0]>, roots: [0], distance: "euclidean" }
-    updated_item_ids: RoaringBitmap<[]>
     "###);
 }
 
 #[test]
 fn write_one_vector_in_one_tree() {
+    let _ = env_logger::builder().is_test(true).parse_filters("trace").try_init();
```
Comment (Kerollmops) [inline src/tests/writer.rs:98]: What is this?
```suggestion
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: src/error.rs [hunk 5]
```diff
@@ -40,10 +42,16 @@ pub enum Error {
...
+    )]
+    MissingMetadata(u16),
+
+    /// The last time items in the database were updated, the [`Writer::build`] method wasn't called.
+    #[error("The trees have not been built after an update on index {0}.")]
```
Comment (Kerollmops) [inline src/error.rs]:
```suggestion
    #[error("The trees have not been built after an update on index {0}")]
```

File: src/error.rs (hunk 1 above)
Comment (irevoire) [reply src/error.rs]: Thanks, done

**Review by Kerollmops — APPROVED** [review]
I love that!

**Pull request merged**